packages/cdktf-cli/lib/dependencies/dependency-manager.ts

```diff
@@ -0,0 +1,102 @@
+import { Language } from "@cdktf/provider-generator";
+import { Errors } from "../errors";
+import { PackageManager } from "./package-manager";
+
+// ref: https://www.terraform.io/language/providers/requirements#source-addresses
+const DEFAULT_HOSTNAME = "registry.terraform.io";
+const DEFAULT_NAMESPACE = "hashicorp";
+function normalizeProviderSource(source: string) {
+  // returns <HOSTNAME>/<NAMESPACE>/<TYPE>
+  const slashes = source.split("/").length - 1;
+  switch (slashes) {
+    case 0:
+      return `${DEFAULT_HOSTNAME}/${DEFAULT_NAMESPACE}/${source}`;
+    case 1:
+      return `${DEFAULT_HOSTNAME}/${source}`;
+    default:
+      return source;
+  }
+}
+
+class ProviderConstraint {
+  public readonly source: string;
+
+  // TODO: parse the version constraint, add examples to cli command description (i.e. =,~>.> etc.)
+  // if no version constraint is specified, we assume the latest version
+  // if specific version is specified without e.g. =, we allow patch level increments (e.g. ~>2.12 for "2.12")
+  constructor(source: string, public readonly version: string | undefined) {
+    this.source = normalizeProviderSource(source);
+  }
+}
+
+export class DependencyManager {
+  private packageManager: PackageManager;
+
+  constructor(
+    private readonly targetLanguage: Language,
+    private cdktfVersion: string
+  ) {
+    this.packageManager = PackageManager.forLanguage(targetLanguage);
+  }
+
+  async addProvider(constraint: ProviderConstraint) {
+    if (await this.hasPrebuiltProvider(constraint)) {
+      return this.addPrebuiltProvider(constraint);
+    } else {
+      return this.addLocalProvider(constraint);
+    }
+  }
+
+  async hasPrebuiltProvider(constraint: ProviderConstraint): Promise<boolean> {
+    // TODO: query NPM for this, add layer that queries this and also caches calls a bit, so we don't have to query NPM every time
+    constraint;
+    // Go has no pre-built providers at the moment
+    if (this.targetLanguage === Language.GO) {
+      return false;
+    }
+
+    return Promise.resolve(false);
+  }
+
+  async addPrebuiltProvider(constraint: ProviderConstraint) {
+    constraint;
+    if (this.targetLanguage === Language.GO) {
+      throw Errors.Usage(
```

Comment:
```suggestion
      throw Errors.Internal(
```

Oder? Du solltest das ja vorher abgechecked haben

Reply:
I thought about a `--pre-built` flag or similar, which might directly invoke `addPrebuiltProvider()` and in which case I would not have caught it before